File: src/db/modules/liveshare/op.py
from sqlalchemy import text

from db.modules.docs.crud import create_document_block, delete_document_block, get_document_block_by_id, get_document_block_by_ids

from db.database import SessionLocal
from sqlalchemy.orm import Session
import uuid

def apply_op_old(content: str, op: dict) -> str:
    if op["type"] == "insert":
        return content[:op["pos"]] + op["text"] + content[op["pos"]:]
    
    if op["type"] == "delete":
        return content[:op["pos"]] + content[op["pos"] + op["length"]:]
    
    return content

class BlockCache:

    def __init__(self, doc_id: int, db: Session):
        self.db = db
        self.doc_id = doc_id
        self.load_index() # lazy load block.id
        self.blocks: dict[str,dict] = {}

        self.new = set()
        self.dirty_content = set()
        self.dirty_position = set()
        self.deleted = set()

    # - db
    def load_index(self):
        rows = self.db.execute(text("""
            SELECT id FROM document_blocks
            WHERE doc_id = :doc_id
            ORDER BY position
        """), {"doc_id": self.doc_id}) # type: ignore

        self.doc_index = [row[0] for row in rows]

    # - system cache
    def load_block(self, block_position: int) -> dict|None:
        if block_position > len(self.doc_index):
            return None
        block_id = self.doc_index[block_position]
        if not self.blocks.get(block_id):
            block = get_document_block_by_id(block_id, self.db)
            if block is None:
                return None
            self.blocks[block_id] = {
                "type": block.type,
                "content": block.content
            }
        return self.blocks[block_id]

    def create_block(self, block_position: int, type: str, content: str) -> bool:
        if block_position > len(self.doc_index):
            return False
        block_id = str(uuid.uuid4())
        self.blocks[block_id] = {
            "type": type,
            "content": content
        }
        self.doc_index.insert(block_position, block_id)
        self.new.update(self.doc_index[block_position])
        self.dirty_position.update(self.doc_index[block_position:])
        return True

    def update_block(self, block_position: int, type: str, content: str) -> bool:
        if block_position > len(self.doc_index):
            return False
        block_id = self.doc_index[block_position]
        block = self.load_block(block_position)
        if block is None:
            return False
        block.update({
            "type": type,
            "content": content
        })
        self.dirty_content.add(block_id)
        return True

    def delete_block(self, block_position: int) -> bool:
        if block_position > len(self.doc_index):
            return False
        block_id = self.doc_index[block_position]
        self.doc_index.pop(block_position)
        self.dirty_position.update(self.doc_index[block_position:])
        self.deleted.add(block_id)
        return True

    # OP
    def apply_op(self, op: dict) -> bool:
        if op["type"] == "create_block":
            return self.create_block(
                op["position"],
                "paragraph",
                ""
            )
        if op["type"] == "update_block":
            return self.update_block(
                op["position"],
                op["type"],
                op["content"]
            )
        if op["type"] == "delete_block":
            return self.delete_block(
                op["position"]
            )
        
        return False
    
    # Flush
    def flush(self):
        # 1. DELETE
        for block_id in self.deleted:
            self.db.execute(
                "DELETE FROM document_blocks WHERE id = :id", # type: ignore
                {"id": block_id}
            ) # type: ignore

        # 2. INSERT (new blocks)
        for block_id in self.new:
            block = self.blocks[block_id]

            position = self.doc_index.index(block_id)

            self.db.execute(text("""
                INSERT INTO document_blocks (id, doc_id, type, content, position)
                VALUES (:id, :doc_id, :type, :content, :position)
            """), { # type: ignore
                "id": block_id,
                "doc_id": self.doc_id,
                "type": block["type"],
                "content": block["content"],
                "position": position
            }) # type: ignore

        # 3. UPDATE content
        for block_id in self.dirty_content:
            if block_id in self.new or block_id in self.deleted:
                continue

            block = self.blocks[block_id]

            self.db.execute(text("""
                UPDATE document_blocks
                SET type = :type, content = :content
                WHERE id = :id
            """), { # type: ignore
                "id": block_id,
                "type": block["type"],
                "content": block["content"]
            }) # type: ignore

        # 4. UPDATE positions
        for position, block_id in enumerate(self.doc_index):
            if block_id in self.deleted:
                continue

            if block_id in self.new or block_id in self.dirty_position:
                self.db.execute(text("""
                    UPDATE document_blocks
                    SET position = :position
                    WHERE id = :id
                """), { # type: ignore
                    "id": block_id,
                    "position": position
                }) # type: ignore

        self.db.commit()

        # 5. clear state
        self.new.clear()
        self.dirty_content.clear()
        self.dirty_position.clear()
        self.deleted.clear()

File: app/src/app/docs/d/[id]/op.ts
import { DocumentBlock } from "@/api/docs";

export type CharOp = 
  | { type: "insert", index: number, text: string }
  | { type: "delete", index: number, length: number }
  | { type: "replace", index: number, text: string, length: number }

export type Op =
  | { type: "insert"; pos: number; text: string }
  | { type: "delete"; pos: number; length: number }
  | { type: "init"; content: DocumentBlock[]; }

  | { type: "create_block"; position: number; id: string }
  | { type: "update_block"; position: number; block_type: string; diff: CharOp }
  | { type: "delete_block"; position: number }

export const applyOp = (val: string, op: any) => {
  if (op.type === "insert") {
    return val.slice(0, op.pos) + op.text + val.slice(op.pos)
  }

  if (op.type === "delete") {
    return val.slice(0, op.pos) + val.slice(op.pos + op.length)
  }

  return val
}

const applyCharOp = (val: string, op: CharOp) => {
  if (op.type == "insert") {
    return val.slice(0, op.index) + op.text + val.slice(op.index)
  }
  if (op.type == "delete") {
    return val.slice(0, op.index) + val.slice(op.index + op.length)
  }
  if (op.type == "replace") {
    return val.slice(0, op.index) + op.text + val.slice(op.index + op.length)
  }
  return val
}

export type Block =
  | {
    id: string;
    type: "paragraph";
    content: string;
  }
  | {
    id: string;
    type: "math_block";
    latex: string;
  };

export type DocumentBlock2 = {
  id: string;
  doc_id?: number;
  position?: number;
  type: string;
  content: string;
  updated_at?: string;
}


type DocumentState = {
  order: string[]                         // [block_id, block_id, ...]
  blocks: Record<string, DocumentBlock2>   // block_id -> block
}


export function reducer(state: DocumentState, op: Op): DocumentState {
  const newState = {
    order: [...state.order],
    blocks: { ...state.blocks }
  }

  switch (op.type) {
    case "init": {
      const blocksArray = op.content

      const order: string[] = []
      const blocks: Record<string, DocumentBlock2> = {}

      for (const block of blocksArray) {
        order.push(block.id)
        blocks[block.id] = block
      }

      return {
        order,
        blocks
      }
    }

    case "create_block": {
      const id = op.id

      newState.order.splice(op.position, 0, id)
      newState.blocks[id] = {
        id,
        type: "paragraph",
        content: ""
      }
      return newState
    }

    case "update_block": {
      const id = newState.order[op.position]
      if (!id) return state

      newState.blocks[id] = {
        ...newState.blocks[id],
        type: op.block_type,
        content: applyCharOp(newState.blocks[id].content, op.diff)
      }
      console.log({newState})
      return newState
    }

    case "delete_block": {
      const id = newState.order[op.position]
      if (!id) return state

      newState.order.splice(op.position, 1)
      delete newState.blocks[id]
      return newState
    }

    default:
      return state
  }
}